beacon_node/eth2_libp2p/src/discovery/mod.rs

```diff
@@ -519,69 +549,116 @@ impl<TSpec: EthSpec> Discovery<TSpec> {
+                            }
+                            Ok(r) => {
+                                let enr: Vec<Enr> = r.clone();
+                                debug!(self.log, "Peer subnet discovery request completed"; "peers_found" => r.len(), "subnet_id" => *subnet_query.subnet_id);
+                                // A subnet query has completed. Add back to the queue, incrementing retries.
+                                self.add_subnet_query(subnet_query.subnet_id, subnet_query.min_ttl, subnet_query.retries + 1);
+
+                                // check specific subnet against enr
+                                let subnet_predicate = subnet_predicate::<TSpec>(vec![subnet_query.subnet_id], &self.log);
+                                let result_enrs: Vec<Enr> = enr.into_iter().filter(subnet_predicate).collect();
+
+                                results.push((subnet_query.min_ttl, result_enrs));
+                            }
+                            Err(e) => {
+                                warn!(self.log,"Subnet Discovery query failed"; "subnet_id" => *subnet_query.subnet_id, "error" => e.to_string());
+                            }
                         }
                     }
                 }
-                QueryType::Subnet {
-                    subnet_id,
-                    min_ttl,
-                    retries,
-                } => {
-                    match query_future.1 {
-                        Ok(r) if r.is_empty() => {
-                            debug!(self.log, "Subnet discovery query yielded no results."; "subnet_id" => *subnet_id, "retries" => retries);
-                        }
-                        Ok(r) => {
-                            debug!(self.log, "Peer subnet discovery request completed"; "peers_found" => r.len(), "subnet_id" => *subnet_id);
-                            // A subnet query has completed. Add back to the queue, incrementing retries.
-                            self.add_subnet_query(subnet_id, min_ttl, retries + 1);
-                            // Report the results back to the peer manager.
-                            return Some((query_future.0.min_ttl(), r));
-                        }
-                        Err(e) => {
-                            warn!(self.log,"Subnet Discovery query failed"; "subnet_id" => *subnet_id, "error" => e.to_string());
+            });
+        }
+
+        let mapped_results = self.map_peers_to_min_ttl(results);
+
+        if mapped_results.is_empty() {
+            return None;
+        } else {
+            Some(mapped_results)
+        }
+    }
+
+    fn map_peers_to_min_ttl(
+        &mut self,
+        results: Vec<(Option<Instant>, Vec<Enr>)>,
+    ) -> HashMap<PeerId, Option<Instant>> {
+        // output mapping
+        let mut enrs_to_min_ttl: HashMap<PeerId, Option<Instant>> = HashMap::new();
+
+        for result in results.iter() {
```

Comment:
Can we consume this? Use this as an `into_iter()` and then we can consume the Vec<Enr>, i.e `result.1.into_iter()`  and remove the `cloned()` and the `enr.clone()`